src/components/transmuters/EthTransmuterLooper.tsx

```diff
@@ -0,0 +1,652 @@
+            : redeemError.message,
+      });
+      return;
+    }
+
+    // 2. Withdraw
+    if (redeemConfig) {
+      redeem(redeemConfig.request);
+    } else {
+      toast.error("Error withdrawing alETH", {
+        description: "Unexpected error. Please contact Alchemix team.",
+      });
+    }
+
+    // TODO -- do I need to refetch the shares/assets balance here, or is it taken care of by the transaction occuring and a refetch in the next block?
+  };
+
+  // Reuses contract logic of _convertToShares for a front end estimation so as not to trigger many RPC calls
+  const previewDeposit = () => {
+    // Get inputted assetsToDeposit, totalAssets, and totalSupply
+    const assetsToDeposit = parseEther(amount); // TODO -- need to validate this input otherwise an error will be thrown. Does token input do this?
+    const totalAssets =
+      (transmuterLooperContractDynamicState?.[
+        TL_DYNAMIC_STATE_INDICIES.totalAssets
+      ] as bigint) || BigInt(0);
+    const totalSupply =
+      (transmuterLooperContractDynamicState?.[
+        TL_DYNAMIC_STATE_INDICIES.totalSupply
+      ] as bigint) || BigInt(0);
+
+    // If assets are 0 but supply is not PPS = 0.
+    if (totalAssets === BigInt(0)) {
+      return totalSupply === BigInt(0) ? formatEther(assetsToDeposit) : 0;
+    }
+
+    // Perform multiplication and division:
+    // assets.mulDiv(totalSupply_, totalAssets_, _rounding);
+    // js truncation handles _rounding = down in the contract call.
+    const expectedShares = (assetsToDeposit * totalSupply) / totalAssets;
+    return formatEther(expectedShares);
+  };
+
+  // Reuses contract logic of _convertToAssets for a front end estimation so as not to trigger many RPC calls
+  const previewRedeem = () => {
+    // Get inputted sharesToRedeem, totalAssets, and totalSupply
+    const sharesToRedeem = parseEther(amount); // TODO -- need to validate this input otherwise an error will be thrown. Does token input do this?
+    const totalAssets =
+      (transmuterLooperContractDynamicState?.[
+        TL_DYNAMIC_STATE_INDICIES.totalAssets
+      ] as bigint) || BigInt(0);
+    const totalSupply =
+      (transmuterLooperContractDynamicState?.[
+        TL_DYNAMIC_STATE_INDICIES.totalSupply
+      ] as bigint) || BigInt(0);
+
+    if (totalSupply == BigInt(0)) {
+      return totalSupply == BigInt(0) ? formatEther(sharesToRedeem) : 0;
+    }
+
+    // Perform multiplication and division:
+    // shares.mulDiv(_totalAssets(S), supply, _rounding);
+    // js truncation handles _rounding = down in the contract call.
+    const expectedAssets = (sharesToRedeem * totalAssets) / totalSupply;
+    return formatEther(expectedAssets);
+  };
+
+  // Calculates and sets previewedOutput for a deposit or withdraw for display purposes.
+  // This represents the previewed output in shares or assets of a deposit or withdraw
+  const previewedOutput = isWithdraw
+    ? (previewRedeem() as string)
+    : (previewDeposit() as string);
```

Comment:
GOOD! no asserting needed, just decide in function which type you want to return.